core/state_transition.go

```diff
@@ -504,19 +504,25 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 		// are 0. This avoids a negative effectiveTip being applied to
 		// the coinbase when simulating calls.
 	} else if kromaConfig != nil {
-		gasUsed := new(big.Int).SetUint64(st.gasUsed())
-		feeDist := st.evm.Context.FeeDistributionFunc(blockNum, gasUsed, st.evm.Context.BaseFee, effectiveTip)
-		st.state.AddBalance(params.KromaValidatorRewardVault, feeDist.Reward)
-		st.state.AddBalance(params.KromaProtocolVault, feeDist.Protocol)
+		if st.evm.ChainConfig().IsKromaMPT(st.evm.Context.Time) {
+			effectiveFee := new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), effectiveTip)
+			st.state.AddBalance(st.evm.Context.Coinbase, effectiveFee)
+			baseFee := new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.evm.Context.BaseFee)
+			st.state.AddBalance(params.KromaProtocolVault, baseFee)
+		} else {
+			feeDist := st.evm.Context.FeeDistributionFunc(blockNum, st.gasUsed(), st.evm.Context.BaseFee, effectiveTip)
+			st.state.AddBalance(params.KromaValidatorRewardVault, feeDist.Reward)
+			st.state.AddBalance(params.KromaProtocolVault, feeDist.Protocol)
+		}
 	} else {
 		fee := new(big.Int).SetUint64(st.gasUsed())
 		fee.Mul(fee, effectiveTip)
 		st.state.AddBalance(st.evm.Context.Coinbase, fee)
 	}
 
 	if kromaConfig != nil && !st.msg.IsDepositTx {
-		if cost := st.evm.Context.L1CostFunc(st.msg.RollupCostData, st.evm.Context.Time); cost != nil {
-			st.state.AddBalance(params.KromaProposerRewardVault, cost)
+		if l1Cost := st.evm.Context.L1CostFunc(st.msg.RollupCostData, st.evm.Context.Time); l1Cost != nil {
+			st.state.AddBalance(params.KromaProposerRewardVault, l1Cost)
 		}
 	}
```

Comment:
To be more compatible with OP and improve readability, I think below version would be better for L499-L527.

```suggestion
	if st.evm.Config.NoBaseFee && msg.GasFeeCap.Sign() == 0 && msg.GasTipCap.Sign() == 0 {
		// Skip fee payment when NoBaseFee is set and the fee fields
		// are 0. This avoids a negative effectiveTip being applied to
		// the coinbase when simulating calls.
	} else if st.evm.ChainConfig().IsPreKromaMPT(st.evm.Context.Time) {
		blockNum := st.evm.Context.BlockNumber.Uint64()
		feeDist := st.evm.Context.FeeDistributionFunc(blockNum, st.gasUsed(), st.evm.Context.BaseFee, effectiveTip)
		st.state.AddBalance(params.KromaValidatorRewardVault, feeDist.Reward)
		st.state.AddBalance(params.KromaProtocolVault, feeDist.Protocol)
	} else {
		fee := new(big.Int).SetUint64(st.gasUsed())
		fee.Mul(fee, effectiveTip)
		st.state.AddBalance(st.evm.Context.Coinbase, fee)
	}

	if kromaConfig := st.evm.ChainConfig().Kroma; kromaConfig != nil && !st.msg.IsDepositTx {
		st.state.AddBalance(params.KromaProtocolVault, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.evm.Context.BaseFee))
		if cost := st.evm.Context.L1CostFunc(st.msg.RollupCostData, st.evm.Context.Time); cost != nil {
			st.state.AddBalance(params.KromaProposerRewardVault, cost)
		}
	}
```